File: mpga-plugin/cli/src/commands/health.ts
import fs from 'fs';
import path from 'path';
import { Command } from 'commander';
import chalk from 'chalk';
import { log, progressBar, miniBanner, gradeColor, statusBadge } from '../core/logger.js';
import { findProjectRoot, loadConfig } from '../core/config.js';
import { runDriftCheck } from '../evidence/drift.js';
import { loadBoard, recalcStats } from '../board/board.js';

export function registerHealth(program: Command): void {
  program
    .command('health')
    .description('Overall project health report')
    .option('--verbose', 'Detailed breakdown')
    .option('--json', 'Machine-readable output')
    .action(async (opts) => {
      const projectRoot = findProjectRoot() ?? process.cwd();
      const mpgaDir = path.join(projectRoot, 'MPGA');
      const config = loadConfig(projectRoot);

      if (!fs.existsSync(mpgaDir)) {
        log.error('MPGA not initialized. Run `mpga init` first.');
        process.exit(1);
      }

      // Gather evidence health
      const driftReport = await runDriftCheck(projectRoot, config.drift.ciThreshold);

      // Gather board health
      const boardDir = path.join(mpgaDir, 'board');
      const tasksDir = path.join(boardDir, 'tasks');
      let boardStats: any = null;
      if (fs.existsSync(path.join(boardDir, 'board.json'))) {
        const board = loadBoard(boardDir);
        recalcStats(board, tasksDir);
        boardStats = board.stats;
      }

      // Gather scope count
      const scopesDir = path.join(mpgaDir, 'scopes');
      const scopeCount = fs.existsSync(scopesDir)
        ? fs.readdirSync(scopesDir).filter(f => f.endsWith('.md')).length
        : 0;

      const health = {
        initialized: true,
        evidenceHealth: driftReport.overallHealthPct,
        evidenceTarget: config.evidence.coverageThreshold * 100,
        ciThreshold: config.drift.ciThreshold,
        ciPass: driftReport.ciPass,
        scopes: scopeCount,
        board: boardStats,
        lastSync: getLastSync(mpgaDir),
        overallGrade: computeGrade(driftReport.overallHealthPct, config.drift.ciThreshold),
      };

      if (opts.json) {
        console.log(JSON.stringify(health, null, 2));
        return;
      }

      miniBanner();
      log.header('Health Report');

      // ── Grade ──
      console.log(`\n  ${chalk.dim('Grade')}  ${gradeColor(health.overallGrade)}\n`);

      // ── Evidence ──
      console.log(`  ${statusBadge(driftReport.overallHealthPct >= 80, 'Evidence health')}   ${driftReport.overallHealthPct}%  ${chalk.dim(`(CI threshold: ${config.drift.ciThreshold}%)`)}`);
      console.log(`    ${progressBar(driftReport.validLinks, driftReport.totalLinks)}  ${chalk.dim(`${driftReport.validLinks}/${driftReport.totalLinks} links`)}`);

      if (opts.verbose && driftReport.scopes.length > 0) {
        log.blank();
        for (const scope of driftReport.scopes) {
          const icon = scope.healthPct >= 80 ? chalk.green('✓') : chalk.yellow('⚠');
          console.log(`    ${icon} ${chalk.white(scope.scope.padEnd(20))} ${scope.healthPct}% ${chalk.dim(`(${scope.validLinks}/${scope.totalLinks})`)}`);
        }
      }

      // ── Scopes ──
      log.blank();
      console.log(`  ${statusBadge(scopeCount > 0, 'Scopes')}            ${scopeCount} document(s)`);

      // ── Board ──
      if (boardStats) {
        log.blank();
        console.log(`  ${statusBadge(boardStats.blocked === 0, 'Task board')}        ${boardStats.done}/${boardStats.total} tasks ${chalk.dim(`(${boardStats.progress_pct}%)`)}`);
        if (boardStats.blocked > 0) console.log(`    ${chalk.yellow('⚠')} ${boardStats.blocked} blocked task(s)`);
      }

      // ── Last sync ──
      log.blank();
      console.log(`  ${chalk.blue('ℹ')} Last sync          ${chalk.dim(health.lastSync)}`);

      // ── Recommendations ──
      log.blank();
      log.divider();
      if (driftReport.overallHealthPct < config.drift.ciThreshold) {
        log.warn(`Evidence below CI threshold — run \`mpga evidence heal\` or \`mpga sync\``);
      }
      if (scopeCount === 0) {
        log.warn('No scope documents — run `mpga sync` to generate them');
      }
      if (!driftReport.ciPass) {
        log.error(`CI would FAIL at ${config.drift.ciThreshold}% threshold`);
      } else {
        log.success('All health checks pass');
      }
      log.blank();
    });
}

function computeGrade(healthPct: number, threshold: number): string {
  if (healthPct >= 95) return 'A';
  if (healthPct >= threshold) return 'B';
  if (healthPct >= threshold * 0.7) return 'C';
  return 'D';
}

function getLastSync(mpgaDir: string): string {
  const indexPath = path.join(mpgaDir, 'INDEX.md');
  if (!fs.existsSync(indexPath)) return 'never';
  const content = fs.readFileSync(indexPath, 'utf-8');
  const m = content.match(/\*\*Last sync:\*\* (.+)/);
  if (m && !m[1].includes('run')) return m[1].trim();
  return 'never';
}

File: mpga-plugin/cli/src/mpga/commands/health.py
from __future__ import annotations

import json
import sys
import time
from pathlib import Path

import click

from mpga.core.config import find_project_root, load_config
from mpga.core.logger import (
    grade_color,
    log,
    mini_banner,
    progress_bar,
    status_badge,
)
from mpga.db.connection import get_connection
from mpga.db.repos.scopes import ScopeRepo
from mpga.db.repos.tasks import TaskRepo
from mpga.db.schema import create_schema
from mpga.evidence.drift import run_drift_check

# Evidence health percentage at or above which status is considered good.
EVIDENCE_HEALTH_GOOD_PCT = 80
# Health percentage at or above which the project earns an A grade.
GRADE_A_THRESHOLD = 95
# Multiplier applied to CI threshold to determine the C grade cutoff.
GRADE_C_MULTIPLIER = 0.7
# TTL for the file-based link-validation cache (seconds).  5 minutes = 300 s.
HEALTH_CACHE_TTL_SECONDS = 300


def _summarize_tasks(tasks) -> dict[str, int]:
    total = len(tasks)
    done = sum(1 for task in tasks if task.column == "done")
    in_flight = sum(1 for task in tasks if task.column in ("in-progress", "testing", "review"))
    blocked = sum(1 for task in tasks if task.status == "blocked")
    progress_pct = round((done / total) * 100) if total else 0
    return {
        "total": total,
        "done": done,
        "in_flight": in_flight,
        "blocked": blocked,
        "progress_pct": progress_pct,
    }


def _load_sqlite_health(project_root: Path) -> dict[str, object] | None:
    project_root = Path(project_root)
    db_path = project_root / ".mpga" / "mpga.db"
    if not db_path.exists():
        return None

    conn = get_connection(str(db_path))
    try:
        create_schema(conn)
        tasks = TaskRepo(conn).filter()
        scopes = ScopeRepo(conn).list_all()
        if not tasks and not scopes:
            return None

        board_stats = _summarize_tasks(tasks) if tasks else None
        last_scanned = conn.execute("SELECT MAX(last_scanned) FROM file_info").fetchone()
        last_sync = last_scanned[0] if last_scanned and last_scanned[0] else "never"
        return {
            "board": board_stats,
            "scopes": len(scopes),
            "last_sync": last_sync,
        }
    finally:
        conn.close()


def _compute_grade(health_pct: int, threshold: int) -> str:
    if health_pct >= GRADE_A_THRESHOLD:
        return "A"
    if health_pct >= threshold:
        return "B"
    if health_pct >= threshold * GRADE_C_MULTIPLIER:
        return "C"
    return "D"


# ---------------------------------------------------------------------------
# Cache helpers — avoid re-scanning all evidence links on every run
# ---------------------------------------------------------------------------

def _cache_path(project_root: Path) -> Path:
    return project_root / ".mpga" / "health_cache.json"


def _read_cache(project_root: Path) -> dict | None:
    """Return cached drift data if the cache exists and is within TTL, else None."""
    path = _cache_path(project_root)
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        age = time.time() - float(data.get("timestamp", 0))
        if age < HEALTH_CACHE_TTL_SECONDS:
            return data
    except (json.JSONDecodeError, ValueError, OSError):
        pass
    return None


def _write_cache(project_root: Path, drift_report) -> None:
    """Persist drift data to the file-based cache with a current timestamp."""
    path = _cache_path(project_root)
    payload = {
        "timestamp": time.time(),
        "overall_health_pct": drift_report.overall_health_pct,
        "valid_links": drift_report.valid_links,
        "total_links": drift_report.total_links,
        "ci_pass": drift_report.ci_pass,
        "scopes": [],  # raw scope objects are not JSON-serialisable; omit for now
    }
    try:
        path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    except (OSError, ValueError):
        pass  # cache write failure is non-fatal


@click.command("health")
@click.option("--verbose", is_flag=True, help="Detailed breakdown")
@click.option("--json", "as_json", is_flag=True, help="Machine-readable output")
def health_cmd(verbose: bool, as_json: bool) -> None:
    """Overall project health report."""
    project_root = Path(find_project_root() or Path.cwd())
    config = load_config(project_root)

    db_exists = (Path(project_root) / ".mpga" / "mpga.db").exists()
    if not db_exists:
        log.error("No MPGA database found. Run `mpga init` first.")
        sys.exit(1)

    sqlite_health = _load_sqlite_health(project_root)

    # Gather evidence health — use file-based cache when still fresh
    cached = _read_cache(project_root)
    if cached is not None:
        # Re-hydrate a lightweight stand-in from the cache so the rest of the
        # function can treat it identically to a real DriftReport.
        from types import SimpleNamespace
        drift_report = SimpleNamespace(
            overall_health_pct=cached["overall_health_pct"],
            valid_links=cached["valid_links"],
            total_links=cached["total_links"],
            ci_pass=cached["ci_pass"],
            scopes=cached.get("scopes", []),
        )
    else:
        drift_report = run_drift_check(str(project_root), config.drift.ci_threshold)
        _write_cache(project_root, drift_report)

    # Gather board health
    board_stats: dict | None = None
    if sqlite_health and sqlite_health["board"]:
        board_stats = sqlite_health["board"]  # type: ignore[assignment]

    # Gather scope count
    scope_count = 0
    if sqlite_health and sqlite_health["scopes"] is not None:
        scope_count = int(sqlite_health["scopes"])

    overall_grade = _compute_grade(drift_report.overall_health_pct, config.drift.ci_threshold)

    health = {
        "initialized": True,
        "evidenceHealth": drift_report.overall_health_pct,
        "evidenceTarget": config.evidence.coverage_threshold * 100,
        "ciThreshold": config.drift.ci_threshold,
        "ciPass": drift_report.ci_pass,
        "scopes": scope_count,
        "board": board_stats,
        "lastSync": str(sqlite_health["last_sync"]) if sqlite_health else "never",
        "overallGrade": overall_grade,
    }

    if as_json:
        click.echo(json.dumps(health, indent=2))
        return

    mini_banner()
    log.header("Health Report \u2014 We're CRUSHING IT")

    # -- Grade --
    click.echo(f"\n  [dim]Grade[/]  {grade_color(overall_grade)}\n")

    # -- Evidence --
    evidence_ok = drift_report.overall_health_pct >= EVIDENCE_HEALTH_GOOD_PCT
    click.echo(
        f"  {status_badge(evidence_ok, 'Evidence health')}   "
        f"{drift_report.overall_health_pct}%  "
        f"[dim](CI threshold: {config.drift.ci_threshold}%)[/]"
    )
    click.echo(
        f"    {progress_bar(drift_report.valid_links, drift_report.total_links)}  "
        f"[dim]{drift_report.valid_links}/{drift_report.total_links} links[/]"
    )

    if verbose and drift_report.scopes:
        log.blank()
        for scope in drift_report.scopes:
            icon = (
                "[green]\u2713[/]"
                if scope.health_pct >= EVIDENCE_HEALTH_GOOD_PCT
                else "[yellow]\u26a0[/]"
            )
            click.echo(
                f"    {icon} {scope.scope:<20} {scope.health_pct}% "
                f"[dim]({scope.valid_links}/{scope.total_links})[/]"
            )

    # -- Scopes --
    log.blank()
    click.echo(f"  {status_badge(scope_count > 0, 'Scopes')}            {scope_count} document(s)")

    # -- Board --
    if board_stats:
        log.blank()
        click.echo(
            f"  {status_badge(board_stats['blocked'] == 0, 'Task board')}        "
            f"{board_stats['done']}/{board_stats['total']} tasks "
            f"[dim]({board_stats['progress_pct']}%)[/]"
        )
        if board_stats["blocked"] > 0:
            click.echo(f"    [yellow]\u26a0[/] {board_stats['blocked']} blocked task(s)")

    # -- Last sync --
    log.blank()
    click.echo(f"  [blue]\u2139[/] Last sync          [dim]{health['lastSync']}[/]")

    # -- Recommendations --
    log.blank()
    log.divider()
    if drift_report.overall_health_pct < config.drift.ci_threshold:
        log.warn("Evidence below CI threshold \u2014 SAD! Run `mpga evidence heal` or `mpga sync`")
    if scope_count == 0:
        log.warn("No scope documents \u2014 SAD! Run `mpga sync` to generate them")
    if not drift_report.ci_pass:
        log.error(
            f"CI would FAIL at {config.drift.ci_threshold}% threshold \u2014 TOTAL DISASTER!"
        )
    else:
        log.success("All health checks PASS \u2014 TREMENDOUS!")
    log.blank()
